osquery/tables/system/windows/shimcache.cpp

```diff
@@ -0,0 +1,216 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed in accordance with the terms specified in
+ *  the LICENSE file found in the root directory of this source tree.
+ */
+
+#include <osquery/core.h>
+#include <osquery/filesystem/fileops.h>
+#include <osquery/logger.h>
+#include <osquery/tables.h>
+#include <osquery/tables/system/windows/registry.h>
+#include <osquery/tables/system/windows/userassist.h>
+
+#include <boost/algorithm/hex.hpp>
+#include <boost/algorithm/string.hpp>
+#include <boost/algorithm/string/predicate.hpp>
+
+#include <string>
+
+const int WIN8 = 256;
+const int WIN10_PRE_CREATOR = 96;
+const int WIN10_CREATOR = 104;
+const std::string WIN8_START = "80";
+const std::string WIN10_START = "30";
+const std::string WIN10_CREATOR_START = "34";
+const std::string WIN_81_10_SHIMCACHE_DELIMITER = "31307473";
+
+// Shimcache can be in multiple ControlSets (ControlSet001, ControlSet002, etc)
+const std::string SHIMCACHE_CONTROLSET =
+    "HKEY_LOCAL_MACHINE\\SYSTEM\\%ControlSet%\\Control\\Session "
+    "Manager\\AppCompatCache";
+
+struct ShimcacheData {
+  std::string path;
+  long long last_modified;
+  std::string execution_flag;
+};
+
+namespace osquery {
+namespace tables {
+
+auto parseShimcacheData(std::string data,
+                        size_t pos,
+                        std::string token,
+                        std::string execution_flag_exists,
+                        std::string delimter) {
+  ShimcacheData shimcache;
+  std::string path_length = token.substr(16, 4);
+
+  // swap endianess
+  std::reverse(path_length.begin(), path_length.end());
+
+  for (std::size_t i = 0; i < path_length.length(); i += 2) {
+    char temp = path_length[i];
+    path_length[i] = path_length[i + 1];
+    path_length[i + 1] = temp;
+  }
+  int shimcache_file_path = std::stoi(path_length, nullptr, 16);
+
+  // Registry data is in Unicode (extra 0x00)
+  std::string path = token.substr(20, shimcache_file_path * 2);
+  boost::erase_all(path, "00");
+  // Windows Store entries have extra data, the extra data includes tabs in the
+  // entry (Unicode value 09). Replace any with spaces due to table formatting
+  // issues
+  boost::replace_all(path, "09", "20");
+  std::string string_path;
+
+  // Convert hex path to readable string
+  try {
+    string_path = boost::algorithm::unhex(path);
+  } catch (boost::algorithm::hex_decode_error) {
```

Comment:
Prefer to catch by const reference: `catch (const boost::algorithm::hex_decode_error&)`, this avoids making a copy of the exception.

Reply:
Fixed